### PR TITLE
Check type name that MUST use keys that match the regular expression: ^[a-zA-Z0-9.-_]+$. 

### PR DESCRIPTION
**Pull request by skywing918**
fix https://github.com/microsoft/typespec/issues/3889

### CODE REVIEW

File: packages/openapi3/src/openapi.ts [hunk 1]
```diff
@@ -1540,6 +1540,21 @@ function createOAPIEmitter(
     }
   }
 
+  function checkAvailableComponentName(type: Type, name: string) {
+    const pattern = /^[\w\\.\\-]+$/;
```
[the remark this answers is not in the document]
Comment (skywing918) [reply packages/openapi3/src/openapi.ts:1544]: [github-advanced-security](https://github.com/apps/github-advanced-security)
`the warning` pop up, so change it...
Code scanning

Overly permissive regular expression range
Medium

Suspicious character range that overlaps with 0-9 in the same character class, and overlaps with A-Z in the same character class, and is equivalent to \[.\\/0-9:;<=>?@A-Z\\\[\\\\\\\]^_\].

[the remark this answers is not in the document]
Comment (wanlwanl) [reply packages/openapi3/src/openapi.ts:1544]: try `/^[a-zA-Z0-9._-]+$/;`

[the remark this answers is not in the document]
Comment (wanlwanl) [reply packages/openapi3/src/openapi.ts:1544]: try `^[a-zA-Z0-9\.\-_]+$`

File: packages/openapi3/src/lib.ts [hunk 2]
```diff
@@ -275,6 +275,12 @@ export const libDef = {
         default: paramMessage`XML \`@unwrapped\` can only used on array properties or primitive ones in the OpenAPI 3 emitter, Property '${"name"}' will be ignored.`,
       },
     },
+    "invalid-type-name": {
+      severity: "error",
+      messages: {
+        default: paramMessage`Invalid component name: '${"value"}'. Component names can only contain the characters 'A-Z a-z 0-9 - . _'.`,
```
[the remark this answers is not in the document]
Comment (wanlwanl) [reply packages/openapi3/src/lib.ts]: please update the pattern:
`[a-zA-Z0-9\.\-_]+$`

File: packages/openapi3/src/openapi.ts [hunk 3]
```diff
@@ -1540,6 +1540,21 @@ function createOAPIEmitter(
     }
   }
 
+  function validateComponentFixedFieldKey(type: Type, name: string) {
+    const pattern = /^[a-zA-Z0-9\\.\-\\_]+$/;
```
Comment (wanlwanl) [inline packages/openapi3/src/openapi.ts:1544]: this allows `\`,
should be `const pattern = /^[a-zA-Z0-9\.\-_]+$/`

Comment (skywing918) [reply packages/openapi3/src/openapi.ts]: hit the Error:   1544:33  error  Unnecessary escape character: \.  no-useless-escape

File: packages/openapi3/src/lib.ts [hunk 4]
```diff
@@ -275,6 +275,12 @@ export const libDef = {
         default: paramMessage`XML \`@unwrapped\` can only used on array properties or primitive ones in the OpenAPI 3 emitter, Property '${"name"}' will be ignored.`,
       },
     },
+    "invalid-component-fixed-field-key": {
+      severity: "error",
+      messages: {
+        default: paramMessage`Invalid key '${"value"}' used in fixed fields of Component object. Only keys matching the regular expression to ^[a-zA-Z0-9.-_]+$ are allowed.`,
```
Comment (wanlwanl) [inline packages/openapi3/src/lib.ts:281]: also update the regex pattern here

File: packages/openapi3/src/openapi.ts [hunk 5]
```diff
@@ -1540,6 +1540,22 @@ function createOAPIEmitter(
     }
   }
 
+  function validateComponentFixedFieldKey(type: Type, name: string) {
+    // eslint-disable-next-line no-useless-escape
+    const pattern = /^[a-zA-Z0-9\.\-_]+$/;
```
Comment (wanlwanl) [inline packages/openapi3/src/openapi.ts:1545]: oh, yes, we don't need `\.`, `/^[a-zA-Z0-9.\-_]+$/` is enough, please use this one, also update the doc. LGTM if it's fixed

**Comment by wanlwanl** [comment]
LGTM

File: packages/openapi3/src/lib.ts [hunk 6]
```diff
@@ -275,6 +275,12 @@ export const libDef = {
         default: paramMessage`XML \`@unwrapped\` can only used on array properties or primitive ones in the OpenAPI 3 emitter, Property '${"name"}' will be ignored.`,
       },
     },
+    "invalid-component-fixed-field-key": {
+      severity: "error",
+      messages: {
+        default: paramMessage`Invalid key '${"value"}' used in fixed fields of Component object. Only keys matching the regular expression to ^[a-zA-Z0-9.\-_]+$ are allowed.`,
```
Comment (markcowl) [inline packages/openapi3/src/lib.ts:281]:
```suggestion
        default: paramMessage`Invalid key '${"value"}' used in a fixed field of the Component object. Only alphanumerics, dot (.), hyphen (-), and underscore (_) characters are allowed in keys.`,
```

**Review by markcowl — CHANGES_REQUESTED** [review]

**Review by markcowl — APPROVED** [review]

**Pull request merged**